### ISSUE TITLE
posts.fav_string is missing favorites

### ISSUE THREAD

**Issue by evazion** [danbooru/danbooru#4652]
In some posts the `fav_string` is missing favorites that are present in the `favorites` table. This happens because we don't lock the post before updating the `fav_string`, so if two users favorite the same post at the same time, one person's favorite can get clobbered and lost. The favorite is still recorded in the favorites table, but it's invisible on the site because it's not present in the `fav_string`.

The fix is to stop using the `fav_string` and treat the favorites table as the single source of truth.

**Comment by evazion** [comment]
You'll have to use the https://danbooru.donmai.us/favorites.json API. There is no pure CSS solution. You can use either https://danbooru.donmai.us/favorites.json?search[post_id]=1 to get the favorites for a single post or https://danbooru.donmai.us/favorites.json?search[post_id]=1,2,3,4,5&search[user_id]=1 to check if a given user has favorited a set of posts.

**Comment by OxxxHayFay** [comment]
Thank you, this is what I need

**Comment by evazion** [comment]
Here are the full gory details of how favorites work (or how they worked up until now) and why this was changed:

* Favorites are stored in two places, in the `favorites` table as a `(id, post_id, user_id)` tuple, and in the posts table as a `fav_string`.
* The `fav_string` is a string of the form `fav:123 fav:456 fav:789`, where the numbers are user IDs.
* The `fav_string` also incidentally tracks the order in which users favorited the post. The first user in the `fav_string` was the first user to favorite the post, the second user was the second favorite, etc. This was used for displaying the list of favorites in the post sidebar in chronological order.
* The `favorites` table also has an ID column that indicates that favorite order. The is used for `ordfav:` searches.
* But for old posts, these orders are inconsistent. The fav order in the `fav_string` differs from the fav order in the `favorites` table for very old posts. Probably because the IDs of the oldest favorites don't appear to be in chronological order. This was probably because of a database migration at some point during Danbooru's early history, which lost the original favorite IDs of the earliest favorites. This are certain other tables too (including the tags and post_versions tables) where early IDs are not in chronological order.
* Unfortunately the `favorites` table doesn't have timestamps, and the `fav_string` is just a flat string, so we have no way to reconcile the true fav ordering when both sources are inconsistent.
* Because of all this, it's hard to fix #2936 (moving favorites breaks favorite ordering). We store the fav order in two different places, and those places may be inconsistent, which makes it really difficult to correctly reconcile the relative fav order between two separate posts.
* The reason *why* the `fav_string` is inconsistent is because updating it is non-atomic. We read the `fav_string` from the database, update it in Ruby, then write it back to the database. This is a non-atomic multi-step process. The post needs to be locked in the database to prevent concurrent updates while the `fav_string` is being updated, otherwise concurrent updates can clobber each other if two people favorite or unfavorite the same post at the exact same time. At Danbooru's scale, this is actually not a rare occurrence.
* A long time ago we had a very pernicious bug where people were could create duplicate favorites if they pressed the fav button too fast. This was also caused by a lack of locking (#1210). The solution was to lock the *user* (in the `users` table) when they favorited a post, but this was wrong. It prevented a single user from favoriting multiple posts at the same time (which solved the duplicate problem), but it didn't prevent multiple users from favoriting the same post at the same time (which allowed the `fav_string` to be clobbered).
* We *could* lock the post instead of the user, which would solve the `fav_string` inconsistency problem, but at this point the damage has already been done. This bug has existed for 10+ years, which means there are thousands of posts with inconsistent `fav_string`s. The best fix is to not denormalize the data, stop using the `fav_string`, and instead use the favorites table as the single source of truth, so we don't have to worry about locking or consistency issues at all.

* The only reason `fav_string` even existed in the first place was just to optimize `fav:<user>` searches. That's it. The `fav_string` effectively treats every favorite as a secret tag on the post. This way `fav:` searches are effectively tag searches. This was done to avoid a JOIN on the `favorites` table for `fav:` searches.
* But we have to do a JOIN on the `favorites` table for `ordfav:` searches *anyway* (to return the results in favorite order), so the `fav_string` hack doesn't help there. It only helps with regular `fav:` searches.
* But it turns out that `ordfav:` searches are much more common than regular `fav:` searches. And even for `fav:` searches, using the `fav_string` over a JOIN only makes a difference if the user has a very large number of favorites. So the `fav_string` doesn't actually help for the majority of favorite searches.
* Even worse, it actually reduced performance of `ordfav:` searches in some cases. It caused the Postgres query planner to choose worse query plans in some cases, especially for users with large numbers of favorites. 
* Because `fav:<user>` searches were secretly tag searches, in the past there were exploits where you could view a user's private favorites by forcing a `fav:<user>` search to be treated as a raw tag (see 2cbe4d36). This is also why we had to remove the raw `fav_string` from the  API; it used to leak private favorites (#3633).

* Because favorites are like secret tags, they get stored in the post's `tag_index` column together with regular tags. This has a bunch of negative consequences:
* About half the `tag_index` is actually devoted to favorites. There are ~100 million favorites on Danbooru and ~130 million tag instances. This means tag searches are slower because the tag index is filled with favorites. It also means Postgres can't collect column statistics as effectively. Postgres tries to keep track of the size of the top 1000 largest tags in the tag index, which helps it with query planning, but because favorites are included in the tag index, a lot of these slots are wasted on random people's favorites instead of real tags.
* It greatly increases database bloat. Because most posts have a lot of favorites, and because favorites are stored inside posts twice (once in the `fav_string`, and once in the `tag_index`), the average size of a post in the database is very high. The average size of a post row is ~1.2kb, of which ~400 bytes are favorites and ~650 bytes are tags. This is a huge amount of bloat. This reduces the performance of everything touching the posts table (which is everything on Danbooru).
* Currently tags are stored in a flat `tag_string`, which sucks. It would better if it were a `tag_ids` integer array. This would make the posts table significantly smaller. But we can't do that as long as the `fav_string` exists, because the `tag_index` is doing double duty containing both tag strings and fav strings (as well as pool strings, but that's another can of worms).
* Updating the `fav_string` causes a lot of write amplification and database churn. Every time a post is favorited, we have to update the `fav_string` on the post, but because Postgres uses MVCC, it doesn't just update the row in place, instead it does a copy-on-write of the entire row. Which means the entire row gets copied and rewritten somewhere else on disk. Which creates fragmentation because posts are constantly moving around. Which is worsened because the average row size is 1.2kb, but the Postgres page size is only 8kb (rows are packed into 8kb pages by Postgres, which means big rows leave big gaps in pages). It also means every index pointing to that post's row has to be updated to point to location of the new row. And because the tag index is an inverted index, if a post has 50 tags and 50 favorites, it basically means 100 indexes to update to point to the new row. All this means: a single favorite triggers a lot of write amplification.

The tl;dr is that:

* The `data-is-favorited` attribute was removed because of internal database changes that mean it can't be included for free anymore.
* Adding it back would add overhead to every search. Admittedly probably not much overhead, but still non-zero. I don't think it's worth it for an invisible html attribute most people don't use.
* There are a *lot* of highly technical reasons for this change, most of which are non-obvious and hard to appreciate without having a deep low-level understanding of both Postgres and Danbooru's internals.

**Comment by BrokenEagle** [comment]
Re: `data-is-favorited` - I added this capability to DisplayPostInfo ([topic #15926](https://danbooru.donmai.us/forum_topics/15926)) with the setting "post favorites enabled".
